src/large_pages/node_large_page.cc

```diff
@@ -144,43 +132,120 @@ inline uintptr_t hugepage_align_down(uintptr_t addr) {
 #endif  // ifndef ElfW
 #endif  // defined(__FreeBSD__)
 
+#define CONSTRUCT_ONLY(class_name)              \
+  class_name(const class_name&) = delete;       \
+  class_name(class_name&&) = delete;            \
+  void operator= (const class_name&) = delete;  \
+  void operator= (const class_name&&) = delete
+
+// Functions in this class must always be inlined because they must end up in
+// the `lpstub` section rather than the `.text` section.
+class MemoryMapPointer {
+ public:
+  CONSTRUCT_ONLY(MemoryMapPointer);
+  FORCE_INLINE explicit MemoryMapPointer() {}
+  FORCE_INLINE bool operator==(void* rhs) const { return mem_ == rhs; }
+  FORCE_INLINE void* mem() const { return mem_; }
+  FORCE_INLINE void Reset(void* start,
+                          size_t size,
+                          int prot,
+                          int flags,
+                          int fd = -1,
+                          size_t offset = 0) {
+    Reset(mmap(start, size, prot, flags, fd, offset), size);
+  }
+  FORCE_INLINE void Reset() { Reset(nullptr, 0); }
+  FORCE_INLINE ~MemoryMapPointer() {
+    if (mem_ == nullptr) return;
+    if (mem_ == MAP_FAILED) return;
+    if (munmap(mem_, size_) == 0) return;
+    PrintSystemError(errno);
+  }
+
+ protected:
+  FORCE_INLINE void Reset(void* data, size_t size) {
+    mem_ = data;
+    size_ = size;
+  }
+
+ private:
+  size_t size_ = 0;
+  void* mem_ = nullptr;
+};
+
+class MappedFilePointer: public MemoryMapPointer {
+ public:
+  CONSTRUCT_ONLY(MappedFilePointer);
+  FORCE_INLINE explicit MappedFilePointer() {}
+  FORCE_INLINE void Reset(void* start,
+                          const char* fname,
+                          int prot,
+                          int flags,
+                          size_t offset = 0) {
+    struct stat file_info;
+    if (stat(fname, &file_info) == -1) goto fail;
+
+    fd_ = open(fname, O_RDONLY);
```

Comment:
`open(2)` is interruptible by a signal. You should retry on EINTR:
```suggestion
    do
      fd_ = open(fname, O_RDONLY);
    while (fd == -1 && errno == EINTR);
```
As well, stat + open is race-y. Not that it really matters here but you still should do open + fstat. :-)